lightning/src/ln/onchaintx.rs

```diff
@@ -0,0 +1,713 @@
+		where B::Target: BroadcasterInterface,
+		      F::Target: FeeEstimator
+	{
+		let mut new_claims = Vec::new();
+		let mut aggregated_claim = HashMap::new();
+		let mut aggregated_soonest = ::std::u32::MAX;
+		let mut new_pending_claim_requests = HashMap::with_capacity(new_claims.len());
+		let mut new_claimable_outpoints = HashMap::with_capacity(claim_requests.len());
+		let mut spendable_outputs = Vec::new();
+
+		// Try to aggregate outputs if they're 1) belong to same parent tx, 2) their
+		// timelock expiration isn't imminent (<= CLTV_SHARED_CLAIM_BUFFER).
+		for (_, siblings_requests) in claim_requests {
+			for req in siblings_requests {
+				// Don't claim a outpoint twice that would be bad for privacy and may uselessly lock a CPFP input for a while
+				if let Some(_) = self.claimable_outpoints.get(&req.outpoint) { log_trace!(self, "Bouncing off outpoint {}:{}, already registered its claiming request", req.outpoint.txid, req.outpoint.vout); } else {
+					log_trace!(self, "Test if outpoint can be aggregated with expiration {} against {}", req.absolute_timelock, height + CLTV_SHARED_CLAIM_BUFFER);
+					if req.absolute_timelock <= height + CLTV_SHARED_CLAIM_BUFFER || !req.aggregable { // Don't aggregate if outpoint absolute timelock is soon or marked as non-aggregable
+						let mut single_input = HashMap::new();
+						single_input.insert(req.outpoint, req.witness_data);
+						new_claims.push((req.absolute_timelock, single_input));
+					} else {
+						aggregated_claim.insert(req.outpoint, req.witness_data);
+						if req.absolute_timelock < aggregated_soonest {
+							aggregated_soonest = req.absolute_timelock;
+						}
+					}
+				}
+			}
+		}
+		new_claims.push((aggregated_soonest, aggregated_claim));
+
+		// Generate claim transactions and track them to bump if necessary at
+		// height timer expiration (i.e in how many blocks we're going to take action).
+		for claim in new_claims {
+			let mut claim_material = ClaimTxBumpMaterial { height_timer: 0, feerate_previous: 0, soonest_timelock: claim.0, per_input_material: claim.1.clone() };
+			if let Some((new_timer, new_feerate, tx)) = self.generate_claim_tx(height, &claim_material, &*fee_estimator) {
+				claim_material.height_timer = new_timer;
+				claim_material.feerate_previous = new_feerate;
+				let txid = tx.txid();
+				new_pending_claim_requests.insert(txid, claim_material);
+				for k in claim.1.keys() {
+					new_claimable_outpoints.insert(k.clone(), (txid, height));
+				}
+				log_trace!(self, "Broadcast onchain {}", log_tx!(tx));
+				spendable_outputs.push(SpendableOutputDescriptor::StaticOutput {
+					outpoint: BitcoinOutPoint { txid: tx.txid(), vout: 0 },
+					output: tx.output[0].clone(),
+				});
+				broadcaster.broadcast_transaction(&tx);
+			}
+		}
+		for (k,v) in new_claimable_outpoints.drain() {
+			log_trace!(self, "Registering claiming request for {}:{}", k.txid, k.vout);
+			self.claimable_outpoints.insert(k, v);
```

Comment:
Why not do this in the loop instead of having another map to track them before adding them later?